File: Problem113/Solution.java
import java.util.*;

public class Solution {
	public List<List<Integer>> pathSum(TreeNode root, int sum) {
		List<List<Integer>> paths = new ArrayList<>();
		getPaths(root, sum, new ArrayList<Integer>(), paths);
		return paths;
	}

	private void getPaths(TreeNode root, int sum, ArrayList<Integer> arrayList, List<List<Integer>> paths) {
		if (root == null)
			return;
		arrayList.add(root.val);
		if (sum == root.val && root.left == null && root.right == null) {
			paths.add(arrayList);
			return;
		}
		getPaths(root.left, sum - root.val, new ArrayList<>(arrayList), paths);
		getPaths(root.right, sum - root.val, new ArrayList<>(arrayList), paths);
	}

	public class TreeNode {
		int val;
		TreeNode left;
		TreeNode right;

		TreeNode(int x) {
			val = x;
		}
	}
}

File: Problem113/Solution.go
package main

type TreeNode struct {
	Val   int
	Left  TreeNode
	Right TreeNode
}

func pathSum(root *TreeNode, sum int) [][]int {
	var result [][]int
	var path []int
	getPath(root, sum, path, &result)
	return result
}

func getPath(root *TreeNode, sum int, path []int, result *[][]int) {
	if root == nil {
		return
	}
	path = append(path, root.Val)
	if sum == root.Val && root.Left == nil && root.Right == nil {
		*result = append(*result, path)
		return
	}
	leftPath := make([]int, len(path))
	RightPath := make([]int, len(path))

	copy(leftPath, path)
	copy(RightPath, path)
	getPath(root.Left, sum-root.Val, leftPath, result)
	getPath(root.Right, sum-root.Val, RightPath, result)
}
